pwndbg/heap/ptmalloc.py

```diff
@@ -5,17 +5,217 @@
+        except:
+            lower = None
+
+        page = None
+        for m in pwndbg.vmmap.get():
+            if m.objfile == '[heap]':
+                page = m
+                break
+
+        if page != None:
+            lower = lower or page.vaddr
+            return (lower, page.vaddr + page.memsz)
+
+        return (None, None)
+
+    def fastbins(self, arena_addr=None):
+        arena        = self.get_arena(arena_addr)
+
+        if arena == None:
+            return
+
+        fastbinsY    = arena['fastbinsY']
+        fd_offset    = self.malloc_chunk.keys().index('fd') * SIZE_SZ
+        num_fastbins = 7
+        size         = SIZE_SZ * 2
+
+        result = OrderedDict()
+        for i in range(num_fastbins):
+            size += SIZE_SZ * 2
+            chain = pwndbg.chain.get(int(fastbinsY[i]), offset=fd_offset)
+
+            result[size] = chain
+
+        return result
+
+    def bin_at(self, index, arena_addr=None):
+        """
+        Modeled after glibc's bin_at function - so starts indexing from 1
+
+        bin_at(1) returns the unsorted bin
+
+        Bin 1          - Unsorted Bin
+        Bin 2 to 63    - Smallbins
+        Bin 64 to 126  - Largebins
+        """
+        assert( index > 0 and index < NBINS )
+        index = index - 1
+
+        arena       = self.get_arena(arena_addr)
+
+        if arena == None:
+            return
+
+        normal_bins = arena['bins']
+        num_bins    = normal_bins.type.sizeof // normal_bins.type.target().sizeof
+
+        bins_base    = int(normal_bins.address) - (SIZE_SZ * 2)
+        current_base = bins_base + (index * SIZE_SZ * 2)
+
+        front, back = normal_bins[index * 2], normal_bins[index * 2 + 1]
+        fd_offset   = self.malloc_chunk.keys().index('fd') * SIZE_SZ
+
+        chain = pwndbg.chain.get(int(front), offset=fd_offset, hard_stop=current_base)
+        return chain
+
+    def unsortedbin(self, arena_addr=None):
+        result = OrderedDict()
+        chain = self.bin_at(1, arena_addr=arena_addr)
+
+        if chain == None:
+            return
+
+        result['all'] = chain
+
+        return result
+
+    def smallbins(self, arena_addr=None):
+        size   = MIN_SMALL_SIZE - (SIZE_SZ * 2)
+        result = OrderedDict()
+
+        for index in range(2, 64):
+            size += SPACES_TABLE[SIZE_SZ][index]
+            chain = self.bin_at(index, arena_addr=arena_addr)
+
+            if chain == None:
+                return
+
+            result[size] = chain
+
+        return result
+
+    def largebins(self, arena_addr=None):
+        size    = MIN_LARGE_SIZE - (SIZE_SZ * 2)
+        result  = OrderedDict()
+
+        for index in range(64, 127):
+            size += SPACES_TABLE[SIZE_SZ][index]
+            chain = self.bin_at(index, arena_addr=arena_addr)
+
+            if chain == None:
+                return
+
+            result[size] = chain
+
+        return result
+
+    def format_bin(self, bins, verbose=False):
```

Comment:
IMHO any display-related code should be in another module.  I don't want to mix display code with logic code, at least where it can be avoided.  I'd suggest `pwndbg.commands.heap` or similar, perhaps a submodule if you like.